### PR TITLE
ipsec: fix stale keys reclaim logic

### PR DESCRIPTION
**Pull request by jibi**
The current logic used to reclaim stale IPSec keys is incorrect. In
fact, whenever a new key is added we just spawn a new goroutine which
after linux_defaults.IPsecKeyDeleteDelay time will delete all keys
different than the one just added.

The issue with this approach is that if a new key is added while one of
these goroutine is still running, this goroutine will end up deleting
the new key.

This commit fixes this by introducing a single new goroutine that
periodically removes any stale key (XFRM state and XFRM policy template)
after enough time has passed since the current key has been added.

Fixes: https://github.com/cilium/cilium/pull/19814

### CODE REVIEW

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 1]
```diff
@@ -710,3 +690,61 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
...
+
+func doReclaimStaleKeys() {
+	ipSecKeysGlobalLock.Lock()
+	defer ipSecKeysGlobalLock.Unlock()
+
+	scopedLog := log.WithField("spi", ipSecCurrentKeySPI)
+
+	xfrmStateList, err := netlink.XfrmStateList(0)
+	if err != nil {
+		scopedLog.WithError(err).Warning("deleting previous SPI, xfrm state list error")
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: If I'm a user and I read this, I'll assume it deleted the previous key and then failed while listing XFRM states. Maybe best to stick to `Failed to list XFRM states`?

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 2]
```diff
@@ -65,10 +66,22 @@ type ipSecKey struct {
 
 var (
 	ipSecKeysGlobalLock lock.RWMutex
```
Comment (kaworu) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Should this mutex be renamed since after this patch it is used to protect more than `ipSecKeysGlobal` (e.g. `doReclaimStaleKeys()` write-lock it without ever accessing `ipSecKeysGlobal`)?

**Review by kaworu — APPROVED** [review]
Thanks @jibi patch LGTM.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 3]
```diff
@@ -710,3 +690,61 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
...
+		// Note that if the key is _not_ in this list we assume it's
+		// there from a previous run of the agent, and so it should be
+		// reclaimed.
+		if ok && time.Now().Sub(markedStaleAt) < linux_defaults.IPsecKeyDeleteDelay {
+			continue
+		}
+
+		scopedLog := scopedLog.WithField("oldSPI", s.Spi)
+
+		scopedLog.Info("New encryption keys reclaiming SPI")
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
		scopedLog.Info("Deleting stale XFRM state.")
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 4]
```diff
@@ -710,3 +690,61 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
...
+		if err := netlink.XfrmStateDel(&s); err != nil {
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:779]: What about the XFRM policies?

**Review by pchaigno — CHANGES_REQUESTED** [review]
I think we also need to clear stale XFRM policies, no?

**Comment by jibi** [comment]
> I think we also need to clear stale XFRM policies, no?

Let me double check, but before this change the goroutine we spawned after a key becomes stale would only delete the XFRM state :thinking: 

https://github.com/cilium/cilium/blob/0b58178dd5d8d9fac123c1cd7bb1c66e7c42c97f/pkg/datapath/linux/ipsec/ipsec_linux.go#L639-L645

https://github.com/cilium/cilium/blob/0b58178dd5d8d9fac123c1cd7bb1c66e7c42c97f/pkg/datapath/linux/ipsec/ipsec_linux.go#L263-L281

**Review by pchaigno — CHANGES_REQUESTED** [review]
What happens if we crash right after loading the new key? Agent will restart, not find anything in the map of stale keys, and remove all XFRM state and policies from previous SPI, right? That will cause connectivity issues with other nodes that have not started using the new SPI yet.

Instead of trying to maintain this state, why can't we rely on the creation timestamps of XFRM states and policies to know if they are older than the timeout? We have those under e.g. `xfrmState.Statistics.AddTime`.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 5]
```diff
@@ -710,3 +700,84 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
...
+		if err != nil {
+			scopedLog.WithError(err).Warning("Failed to list XFRM policies")
+		}
+
+		// After the state, delete also any XFRM policy template
+		// referencing the stale key
+	nextPolicy:
+		for _, p := range xfrmPolicyList {
+			if ipSecPolicyHasTemplateForSPI(p, s.Spi) {
+				if err := xfrmDeleteTemplateFromPolicy(p, s.Spi); err != nil {
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: We want to delete the policy, not just its templates (we always have only one).

The SPI is also part of the XFRM policy's mark, though I'm not sure if it's best to check that or the template's SPI field.

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: pkg/datapath/linux/ipsec/ipsec_linux.go (hunk 5 above)
Comment (jibi) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: I've switched to the XFRM policy's mark test since it simplifies the code and we can assume there's always only one template per policy

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 6]
```diff
@@ -260,24 +265,8 @@ func ipSecReplacePolicyOut(src, dst, tmplSrc, tmplDst *net.IPNet, dir IPSecDir)
...
-	}
-	for _, s := range xfrmStateList {
-		if s.Spi != int(spi) {
-			if err := netlink.XfrmStateDel(&s); err != nil {
-				scopedLog.WithError(err).Warning("deleting old xfrm state failed")
-			}
-		}
-	}
+func ipSecPolicyForSPI(policy netlink.XfrmPolicy, spi int) bool {
+	return int(policy.Mark.Value>>12) == spi
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: We should define this value as a constant at the top. We already use it in two places without a proper constant, but this time it's going to be easier to miss if we ever change the value. The two other places will break connectivity if they have conflicting shift values; here it may just cause XFRM states to leak, which our CI won't surface.

It can probably be a separate PR if you want to avoid another iteration through CI.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 7]
```diff
@@ -710,3 +676,83 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
...
+		// stale key
+		for _, p := range xfrmPolicyList {
+			if !ipSecPolicyForSPI(p, s.Spi) {
+				continue
+			}
+
+			if err := netlink.XfrmPolicyDel(&p); err != nil {
+				scopedLog.WithError(err).Warning("Deleting stale XFRM policy failed")
+			}
+		}
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:811]: You can move this second loop outside the first one if you switch `ipSecPolicyForSPI()` to a parser of the SPI (e.g., `getSPIFromXfrmPolicy()`) and just run through the list of XFRM policies, deleting those with a stale SPI, same as you're doing for XFRM states.

**Review by pchaigno — APPROVED** [review]
Two minor comments below.

**Comment by jibi** [comment]
> Two minor comments below.

Thanks for the review :pray: 

New changes with this iteration:
* defined a new constant for that shift value
* introduced a couple of helpers to set and get the SPI from a XFRM mark
* move the loop for deleting policies outside the state one

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 8]
```diff
@@ -253,33 +265,13 @@ func ipSecReplacePolicyOut(src, dst, tmplSrc, tmplDst *net.IPNet, dir IPSecDir)
...
-		return
-	}
-	for _, s := range xfrmStateList {
-		if s.Spi != int(spi) {
-			if err := netlink.XfrmStateDel(&s); err != nil {
-				scopedLog.WithError(err).Warning("deleting old xfrm state failed")
-			}
-		}
-	}
-}
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:281]: Shouldn't this removal be in a different commit?

Comment (jibi) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:281]: Yes, messed up with rebase, it should be in the subsequent commit

edit: we actually still need it for tests, I'll update `doReclaimStaleKeys()` to use it rather than inlining the logic

Comment (pchaigno) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:281]: `git rebase --exec "make build -j$(nproc)" origin/master` for the win :wink:

**Review by pchaigno — APPROVED** [review]
LGTM! (Nice commit history!)

**Review by twpayne — APPROVED** [review]
👍  for CLI

**Comment by jibi** [comment]
Updated to only take into account OUT XFRM policies when deleting stale ones

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 9]
```diff
@@ -253,29 +278,32 @@ func ipSecReplacePolicyOut(src, dst, tmplSrc, tmplDst *net.IPNet, dir IPSecDir)
 	policy.Dst = &net.IPNet{IP: dst.IP.Mask(dst.Mask), Mask: dst.Mask}
 	policy.Dir = netlink.XFRM_DIR_OUT
 	policy.Mark = &netlink.XfrmMark{
-		Value: ((spiWide << 12) | linux_defaults.RouteMarkEncrypt),
+		Value: ipSecXfrmMarkSetSPI(linux_defaults.RouteMarkToProxy, key.Spi),
```
Comment (pchaigno) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: Typo I guess:
```suggestion
		Value: ipSecXfrmMarkSetSPI(linux_defaults.RouteMarkEncrypt, key.Spi),
```

Comment (jibi) [reply pkg/datapath/linux/ipsec/ipsec_linux.go]: :scream:

**Comment by jibi** [comment]
Since the last review I updated the PR with an additional call to `NodeValidateImplementation()` in `keyfileWatcher()`:

https://github.com/cilium/cilium/blob/76a5bb907c2418050598fde28046fb4396507f9d/pkg/datapath/linux/ipsec/ipsec_linux.go#L729-L733

This should ensure that whenever a key is rotated the relevant XFRM states and policy are updated immediately rather than waiting on the `backgroundSync()` method of the `node/manager` package to eventually update the configuration

**Review by pchaigno — APPROVED** [review]

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 10]
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
...
+	go keyfileWatcher(ctx, watcher, keyfilePath, nodediscovery, nodeHandler)
 
 	return nil
 }
+
+func doReclaimStaleKeys() {
+	ipSecLock.Lock()
+	defer ipSecLock.Unlock()
+
+	xfrmStateList, err := netlink.XfrmStateList(0)
```
Comment (julianwiedmann) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
	xfrmStateList, err := netlink.XfrmStateList(netlink.FAMILY_ALL)
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 11]
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
...
+	// - loadIPSecKeys() will only set ipSecCurrentKeySPI to the the SPI of
+	//   the new key
+	// - the actual IPSec state for the new key will be added when we
+	//   receive the k8s nodes from kube-apiserver and the local node from
+	//   our discovery
+	//
+	// Because of this, initialize currentKeyCreatedAt to the current time
+	// as we assume the current key is still being created unless we find
+	// its state.
+	currentKeyCreatedAt := time.Now()
```
Comment (julianwiedmann) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]: This is fine for now. But I really think we should just break out when the `ipSecCurrentKeySPI` hasn't been installed yet. Avoid any doubt whether the subsequent `time.Now()` arithmetic will catch this in all corner cases.

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 12]
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
...
+		log.WithField(logfields.SPI, ipSecCurrentKeySPI).
+			WithError(err).Warning("Failed to list XFRM states")
+		return
+	}
+
+	// First we need to find when the current key was added by looking at
+	// the AddTime of the IPSec state for the key.
+	//
+	// The creation of the IPSec state is asynchronous to loadIPSecKeys():
+	// - loadIPSecKeys() will only set ipSecCurrentKeySPI to the the SPI of
```
Comment (julianwiedmann) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
	// - loadIPSecKeys() will only set ipSecCurrentKeySPI to the SPI of
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 13]
```diff
@@ -299,6 +328,34 @@ func ipsecDeleteXfrmState(ip net.IP) {
 	}
 }
 
+func ipsecDeleteXfrmPoliciesNotMatchingSPI(spi uint8) {
+	scopedLog := log.WithField(logfields.SPI, spi)
+
+	xfrmPolicyList, err := netlink.XfrmPolicyList(0)
```
Comment (julianwiedmann) [inline pkg/datapath/linux/ipsec/ipsec_linux.go]:
```suggestion
	xfrmPolicyList, err := netlink.XfrmPolicyList(netlink.FAMILY_ALL)
```

File: pkg/datapath/linux/ipsec/ipsec_linux.go [hunk 14]
```diff
@@ -700,13 +746,75 @@ func keyfileWatcher(ctx context.Context, watcher *fswatcher.Watcher, keyfilePath
...
+		}
+
+		currentKeyCreatedAt = time.Unix(int64(s.Statistics.AddTime), 0)
+		break
+	}
+
+	// If the current key was added less then IPSecKeyDeleteDelay time ago,
+	// don't try to reclaim any stale state or policy
+	if time.Now().Sub(currentKeyCreatedAt) < linux_defaults.IPsecKeyDeleteDelay {
+		return
```
Comment (julianwiedmann) [inline pkg/datapath/linux/ipsec/ipsec_linux.go:822]: One (obvious) caveat is that users should not cycle their keys faster than `IPsecKeyDeleteDelay` :).

**Review by julianwiedmann — APPROVED** [review]
lgtm, just some minor comments. Don't think we need to update `Documentation/gettingstarted/encryption-ipsec.rst` either, it's all implementation details.

File: pkg/datapath/linux/ipsec/ipsec_linux.go (hunk 14 above)
Comment (jibi) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:822]: Correct. The initial approach was based on maintaining some state (the creation timestamp of each key) but as Paul pointed out that may cause connectivity issues in some circumstances: https://github.com/cilium/cilium/pull/19932#pullrequestreview-983002135

Maybe we can resort to use both approaches :thinking:

Comment (odinuge) [reply pkg/datapath/linux/ipsec/ipsec_linux.go:822]: I'm also curios about the situation where a new key is inserted to the k8s configmap, and a new node start up immediately. The new node will then only have the new key, while it can take a few minutes until all other nodes have the new key, since this solution will rely on the kubelet updating the configmap on disk.

In such a situation, will a new node be blocked from scheduling pod networked traffic, or will it be able to schedule pods successfully, but they won't have pod network connectivity the first few minutes.. 🤔  Haven't had time to test it yet.

**Comment by jibi** [comment]
Latest iteration, taking @julianwiedmann feedback into account.

We don't want to assume that a user is always rotating keys less frequently than the `linux_defaults.IPsecKeyDeleteDelay` interval, and the current approach based on XFRM state timestamp is not going to work here as we need the time a key was deleted rather than the one it was added.

This new approach reintroduces the `ipSecKeysRemovalTime` local state to keep track of removal timestamps, and if an SPI is not in that map we assume it was just deleted and wait `IPsecKeyDeleteDelay` before removing it (in this way we manage also the case where the agent gets restarted while stale keys are still present, as after enough time has passed since the )

**Review by qmonnet — APPROVED** [review]

**Comment by jibi** [comment]
/test-1.16-4.9

**Comment by jibi** [comment]
/test-1.16-4.9

**Comment by jibi** [comment]
/test-1.24-net-next

**Comment by jibi** [comment]
/test-1.16-4.9

**Comment by julianwiedmann** [comment]
> Latest iteration, taking @julianwiedmann feedback into account.
> 
> This new approach reintroduces the `ipSecKeysRemovalTime` local state to keep track of removal timestamps, and if an SPI is not in that map we assume it was just deleted and wait `IPsecKeyDeleteDelay` before removing it (in this way we manage also the case where the agent gets restarted while stale keys are still present, as after enough time has passed since the )

There's another reason why this is needed. If `doReclaimStaleKeys()` hits the time window just right so that **only** `deleteStaleXfrmPolicies()` reclaims a SPI, then we need to give `deleteStaleXfrmStates()` a second chance for this SPI when the reclaimer runs again.

If we can't avoid this scenario (ie. have a smart approach where we evaluate `ipSecSPICanBeReclaimed()` just **once** for each SPI), then we should at least adjust the comment in `ipSecSPICanBeReclaimed()`.

**Comment by julianwiedmann** [comment]
Been trying to poke holes into the reclaim logic, but I think this should work :+1:. The agent-restart scenarios are certainly ... interesting. With the async filewatcher and reclaimer, I hope we still ensure that all the config happens in the right order.

Note that the commit msg for `ipsec: fix stale keys reclaim logic` might also need an update:

> ...
> This commit fixes this by introducing a single new goroutine that
> periodically removes any stale key (XFRM state and XFRM policy) after
> ~~enough time has passed since the current key has been added.~~
> **linux_defaults.IPsecKeyDeleteDelay has passed since this key was replaced.**

**Comment by jibi** [comment]
> If we can't avoid this scenario (ie. have a smart approach where we evaluate `ipSecSPICanBeReclaimed()` just **once** for each SPI), then we should at least adjust the comment in `ipSecSPICanBeReclaimed()`.

Nice catch! What complicates things a bit is the fact that there are multiple states (4) and policies (2) for any given SPI, so we may as well end up with only half of the states removed :thinking: 

I don't think an approach based on precalculating the list of SPIs (i.e .evaluate `ipSecSPICanBeReclaimed()` just once) could work:
* we cannot assume that all states and policies for a given SPI are either all present or all removed (thinking about the case where the agent is restarted right before starting deleting old policies)
* because of this we would still have to build the list of SPIs that need to be reclaimed by looking both at states and policies, going back to the original problem

Given that the issue is `ipSecSPICanBeReclaimed()` being dependent on the time it is called, what about getting that out of the equation and making that function take a timestamp (which then would be the timestamp of the  `doReclaimStaleKeys()` invocation?)

**Comment by julianwiedmann** [comment]
> > If we can't avoid this scenario (ie. have a smart approach where we evaluate `ipSecSPICanBeReclaimed()` just **once** for each SPI), then we should at least adjust the comment in `ipSecSPICanBeReclaimed()`.
> 
> Nice catch! What complicates things a bit is the fact that there are multiple states (4) and policies (2) for any given SPI, so we may as well end up with only half of the states removed thinking
> 
> I don't think an approach based on precalculating the list of SPIs (i.e .evaluate `ipSecSPICanBeReclaimed()` just once) could work:
> 
>     * we cannot assume that all states and policies for a given SPI are either all present or all removed (thinking about the case where the agent is restarted right before starting deleting old policies)
> 
>     * because of this we would still have to build the list of SPIs that need to be reclaimed by looking both at states and policies, going back to the original problem
> 
Agree, there's a lot of complexity there.

> 
> Given that the issue is `ipSecSPICanBeReclaimed()` being dependent on the time it is called, what about getting that out of the equation and making that function take a timestamp (which then would be the timestamp of the `doReclaimStaleKeys()` invocation?)

Sounds good!

**Review by NikAleksandrov — APPROVED** [review]

**Comment by jibi** [comment]
/test-1.24-net-next
